File: packages/train/src/experiments/common/solveEpisode.ts
import { Bin, Board, moveToString } from '../../game';
import { force } from '../../utils';
import { playEpisode } from './playEpisode';
import type { LayersModel } from './tf';
import type { ProcessOneMoveType } from './types';
import { getLastBinFromEpisode } from './utils';

export const solveEpisode = async (
    model: LayersModel,
    processOneMove: ProcessOneMoveType,
    startBoard: Board,
    expectedBin: Bin,
    stepsLimit: number,
    verbose = false,
): Promise<{
    history: string[];
    finalBin: string[];
}> => {
    const episode = await playEpisode(
        model,
        processOneMove,
        startBoard,
        expectedBin,
        stepsLimit,
        force,
        verbose,
    );
    const finalBin = getLastBinFromEpisode(episode);
    const history = episode.map((step) => moveToString(step.board, step.move));

    return {
        history,
        finalBin,
    };
};

File: pythonFramework/experiments/solve_episode.py
import numpy as np

from .play_episode import playEpisode
from .utils import getLastBinFromEpisode
from ..game.utils import moveToString


def solveEpisode(
        model,
        processOneMove,
        startBoard,
        expectedBin,
        stepsLimit,
        verbose=False
):
    episode = playEpisode(
        model,
        processOneMove,
        startBoard,
        expectedBin,
        stepsLimit,
        np.argmax,
        verbose,
    )
    finalBin = getLastBinFromEpisode(episode)
    history = list(map(lambda step: moveToString(step['board'], step['move']), episode))

    return [history, finalBin]
